### PR TITLE
Feat inventories

### PR DESCRIPTION
**Pull request by apachano**
Resolves: #34 
Authors:
@apachano 

## Summary of changes
- Implements creative and player inventories

## Caveats
Does this introduce any new bugs?
- "held" item shows up in statically placed inv slot instead of replacing cursor

## QA Notes
- Items in creative inventory should be unlimited
- Items in player inventory should be limited
- Only one item should be held at a time
- Only one item should fix in a slot
- First 10 slots in inventory should map to hot bar

## On approval
Merge

### CODE REVIEW

**Comment by tobyplowy** [comment]
@apachano CI build fails.

**Comment by apachano** [comment]
Fixed the build issue

**Comment by tobyplowy** [comment]
Build failed with 1 Error.
[build log).txt](https://github.com/GentenStudios/Phoenix/files/5773063/message.1.txt)

**Comment by apachano** [comment]
@tobyplowy Should be fixed now

**Comment by tobyplowy** [comment]
Builds and runs.
Also this is so COOL.
image
Good stuff so far keep it up!

File: Phoenix/Client/Source/HUD.cpp [hunk 1]
```diff
@@ -70,8 +70,28 @@ void HUD::tick(float dt)
 			ImGui::Text("");
 		}
 
-		ImGui::Text("Hand: %s",
-		            m_registry->get<Hand>(m_player).hand->displayName.c_str());
+		Hand& hand = m_registry->get<Hand>(m_player);
+		for (size_t i = 0; i < hand.size; i++)
```
Comment (beeperdeeper089) [inline Phoenix/Client/Source/HUD.cpp:74]: std::size_t

File: Phoenix/Client/Source/InventoryUI.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+		{
+			ImGui::SameLine();
+		}
+		const voxels::Item item = m_inventory->getItem(i);
+		ImGui::PushID(i);
+		if (item.type == nullptr)
+		{
+			if (ImGui::Button("", {50, 50}))
+			{
+				LOG_DEBUG("Inv") << "Click";
```
Comment (beeperdeeper089) [inline Phoenix/Client/Source/InventoryUI.cpp:100]: Do we still need this debug statement?

File: Phoenix/Common/Include/Common/Voxels/Inventory.hpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+		size_t                        getSize() const { return m_size; };
+
+		// serialize.
+		Serializer& operator>>(Serializer& ser) const override;
+
+		// deserialize.
+		Serializer& operator<<(Serializer& ser) override;
+
+	private:
+		size_t                 m_size;
```
Comment (beeperdeeper089) [inline Phoenix/Common/Include/Common/Voxels/Inventory.hpp:105]: std::size_t

File: Phoenix/Common/Include/Common/Voxels/Inventory.hpp [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		 * @param newData The new value for the data.
+		 * @return true if the data was set.
+		 * @return false if the data already exists with a different data type
+		 * OR if the supplied slot is out of bounds for the inventory.
+		 */
+		bool setMetadataAt(std::size_t slot, const std::string& key,
+		                   std::any* newData);
+
+		const std::vector<ItemType*>& getItems() const { return m_slots; };
+		size_t                        getSize() const { return m_size; };
```
Comment (beeperdeeper089) [inline Phoenix/Common/Include/Common/Voxels/Inventory.hpp:96]: std::size_t

**Review by beeperdeeper089 — CHANGES_REQUESTED** [review]

File: Phoenix/Client/Include/Client/HUD.hpp [hunk 5]
```diff
@@ -54,9 +54,8 @@ namespace phx::client
 		void onEvent(events::Event& e) override;
 
 		void tick(float dt) override;
-
-		static constexpr float WIDTH  = 500;
-		static constexpr float HEIGHT = 50;
+		static constexpr float WIDTH  = 588;
```
[the remark this answers is not in the document]
Comment (apachano) [reply Phoenix/Client/Include/Client/HUD.hpp:57]: Added compile time calculation

File: Phoenix/Client/Source/Game.cpp [hunk 6]
```diff
@@ -123,12 +123,18 @@ void Game::onAttach()
...
 	}
 	else
 	{
 		m_map = new voxels::Map(m_save, "map1", &m_blockRegistry.referrer);
 	}
+	m_invManager =
+	    new voxels::InventoryManager(m_save, &m_itemRegistry.referrer);
+	m_playerInventory =
+	    m_invManager->getInventory(m_invManager->createInventory(30));
+	m_registry->emplace<Hand>(m_player, std::size_t(1), m_playerInventory);
```
[the remark this answers is not in the document]
Comment (apachano) [reply Phoenix/Client/Source/Game.cpp:137]: Apparently I can, I remember hitting a compile error but its gone now

**Review by beeperdeeper089 — APPROVED** [review]
Lgtm, wait for Toby to test and I'm happy to merge.

**Comment by tobyplowy** [comment]
Bulds and runs as intened. Hotbar issue fixed. 
Dud stuff austin my guy.

**Pull request merged**